### ISSUE TITLE
[Bug/Question] `defaultFocusChild` does not work at mount

### ISSUE THREAD

**Issue by LBRDan** [jamesplease/lrud#70]
`defaultFocusChild`, provided in numeric or functional form `() => number`, does not seem to work at component mount time
Using the Codesandbox link provides a reproduction case

Codesandbox: [HERE](https://codesandbox.io/s/determined-greider-pm5kk)
(Sorry, no UI, logs in console only)

## Currently:
```

defaultFocusChild (EXPECTING SECTION_CONTAINER) 
FOCUSED NAV_ITEM_1 
FOCUSED NAV_CONTAINER 
```

## Expected:
```

defaultFocusChild (EXPECTING SECTION_CONTAINER) 
FOCUSED APP 
FOCUSED SECTION_CONTAINER 
```

Also, seems that `onMountAssignFocusTo` works as expected

This appears to be a bug enclosed somewhere between the `createNodes`/`updateFocus` running on an incompletely registered tree.

Q: Is `defaultFocusChild` to be used in update only property?
Thanks 🙏

**Comment by jamesplease** [comment]
It'll be at least a few weeks until I can look into this, sorry!

**Comment by LBRDan** [comment]
No problem!
Is there anything i can do to help you?

**Comment by jamesplease** [comment]
Opening a PR with the fix if there’s a bug would be helpful ✌️

**Comment by LBRDan** [comment]
I should try to do it, but it looks like a design hiccup and I seriously cannot wrap my head around the entire thing...

**Comment by jamesplease** [comment]
Yeah the lib is a bit complicated on the implementation side. I’ll likely be doing a massive rewrite/refactor in 2022 sometime so if you continue opening up issues for problems you encounter (thank you by the way!) I can make sure they get fixed at that time, at the very least.

**Comment by jamesplease** [comment]
You're right that I generally use `onMountAssignFocusTo` for mounting and `defaultFocusChild` for navigation after mounting.

So in the CodeSandbox you provided I would use `onMountAssignFocusTo`.

I plan to write out a guide explaining this in more detail, but for now,  a few differences between the two:

- `onMountAssignFocusTo` allows you to specify a deeply nested child to receive the focus. It's ideal for mounting an entire "page" of nodes and being able to specify that a particular node on that page receive focus -- no matter where it is in the tree.
- `defaultFocusChild` refers only to specific children. It was designed to provide a way to override the default navigation behavior of the lib as the user is LRUD'ing around the app.

I need to think more about the interaction between the two. I'd say it's probably a bug that `defaultFocusChild` isn't being factored in on the initial mount, but I need to look more into the mounting algo to verify that it is possible to factor it in without introducing regressions.

**Comment by jamesplease** [comment]
Some notes for myself:

- The tl;dr of the problem here is that `onMountAssignFocusTo` freezes the focus calculation on mount, while `defaultFocusChild` does not

More details:

- `defaultFocusChild` is used in `tree-navigation`, specifically when finding children
- It is being read correctly here, but clamped to `0`. This is because at the time that it is considered, the other child hasn't been added to the tree
- `onMountAssignFocus` freezes (published) state updates until the final node exists

Question for myself:

- should ALL updates freeze focus? This would require splitting out the `createNode` logic into 2 steps: 1. creation of the tree 2. assigning of focus, starting at the top

That idea makes sense to me rn, but I'll need to dig into the algorithm more to see if there are pitfalls with that approach.

**Comment by jamesplease** [comment]
Because the fix for this (#95) has the potential to be breaking for folks this will be released in the next major version. It may be a bit, however, as I try to avoid breaking changes.

Might bundle it up with #68 and a possible resolution to #92

**Comment by jamesplease** [comment]
Going to close this and just track it over in #98 with the big refactor that I'm starting.
